### PR TITLE
Modelcontrol component

### PR DESCRIPTION
**Pull request by korenkovv**

### CODE REVIEW

File: WebClient/doc/components/moduleControl/moduleControl.md [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+    * [ModuleControlWindowViewController](#moduleControlWindowViewControllerApi)
+    * [ModuleControlToggleViewModel](#moduleControlToggleViewModelApi)
+    * [ModuleControlToggleView](#moduleControlToggleViewApi)
+    * [ModuleControlToggleViewController](#moduleControlToggleViewControllerApi)
+
+## Architecture <a name="architecture"></a>
+
+ModuleControl component consists of several domains:
+
+* *ModuleControlService* provides Modlue models to the application;
```
Comment (AnganaTNO) [inline WebClient/doc/components/moduleControl/moduleControl.md]: Spelling of Module is not correct.

Comment (korenkovv) [reply WebClient/doc/components/moduleControl/moduleControl.md]: 👍 will fix

File: WebClient/doc/core/window/windowManager/windowManager.md [hunk 2]
```diff
@@ -58,3 +58,17 @@ Options object that should be passed to *WindowManager* constructor.
...
+
+#### AddWindowOptions <a name="addWindowOptions"></a>
+
+| Property | Type | Description |
+|---|---|---|
+| position | [WindowPosition](#windowPosition) | Property indicates window position. |
+
+#### WindowPosition <a name="windowPosition"></a>
+
+*WindowPosition* is a string coinstant taht can be one of the following values.
```
Comment (AnganaTNO) [inline WebClient/doc/core/window/windowManager/windowManager.md]: some minor spelling errors  in this line.

File: WebClient/src/components/moduleControl/moduleControlToggle/moduleControlToggleView.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+var ModuleControlToggleView = ControlView.extend({    
+    options: {
+        collapsed: true,
+        position: 'bottomright',
+        autoZIndex: true
+    },
+
+    onInitialize: function (opts) {
+        if (!opts) throw new Error('opts is not provided');
+        if (!opts.toggleViewModel) throw new Error('opts.toggleViewModel is nleaflet-control-model-toggleot provided');
```
Comment (AnganaTNO) [inline WebClient/src/components/moduleControl/moduleControlToggle/moduleControlToggleView.js]: The error message here does not feel correct. Is that so?

File: WebClient/src/components/moduleControl/moduleControlWindow/moduleControlItemList/moduleControlItemListView.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+import './moduleControlItemList.css';
+import View from '../../../../core/view';
+import ModuleControlItemView from '../moduleControlItem/moduleControlItemView';
+
+var ModuleControlItemListView = View.extend({
+
+    onInitialize: function (opts) {
+        if (!opts) throw new Error('No arguments are provided to the View');
+        if (!opts.modules) throw new Error('modules is not provided');
```
Comment (AnganaTNO) [inline WebClient/src/components/moduleControl/moduleControlWindow/moduleControlItemList/moduleControlItemListView.js]: I think it should be 'Modules are not provided'.

File: WebClient/src/components/moduleControl/moduleControlWindow/moduleControlItem/moduleControlItemView.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            .classed('fa-spin', this._moduleControlItemModel.loading)
+            .classed('fa-check', !this._moduleControlItemModel.loading)            
+            .classed('module-status-icon-ready', !this._moduleControlItemModel.loading);
+        
+        // workaround: IE11 and Edge do not redraw an element after class is removed
+        // therefore icon will remain spinning even after fa-spin class is removed.
+        // This can be fixed with changing the element thus forceing element redraweal.
+        if (!this._moduleControlItemModel.loading) {
+            d3.select(this._rootElement).select('.module-status-icon')
+                .style('trarsform', 'none');
```
Comment (AnganaTNO) [inline WebClient/src/components/moduleControl/moduleControlWindow/moduleControlItem/moduleControlItemView.js]: It should be .style( 'transform','none' );'

File: WebClient/doc/core/window/windowManager/windowManager.md (hunk 2 above)
Comment (korenkovv) [reply WebClient/doc/core/window/windowManager/windowManager.md]: 👍

File: WebClient/src/components/moduleControl/moduleControlToggle/moduleControlToggleView.js (hunk 3 above)
Comment (korenkovv) [reply WebClient/src/components/moduleControl/moduleControlToggle/moduleControlToggleView.js]: :)
👍

File: WebClient/src/components/moduleControl/moduleControlWindow/moduleControlItemList/moduleControlItemListView.js (hunk 4 above)
Comment (korenkovv) [reply WebClient/src/components/moduleControl/moduleControlWindow/moduleControlItemList/moduleControlItemListView.js]: Changed to 'opts.modules property is not... '

File: WebClient/src/components/moduleControl/moduleControlWindow/moduleControlItem/moduleControlItemView.js (hunk 5 above)
Comment (korenkovv) [reply WebClient/src/components/moduleControl/moduleControlWindow/moduleControlItem/moduleControlItemView.js]: 👍

**Review by AnganaTNO — APPROVED** [review]
Done.

File: WebClient/src/components/moduleControl/moduleModel.js [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    LOCKED:      'locked',
+    BUSY:        'busy',
+    READY:       'ready',
+    CALCULATING: 'calc',
+    REMOVED:     'removed'
+};
+
+export function validateStatus(status) {
+    for (var s in ModuleStatus) {        
+        if (!ModuleStatus.hasOwnProperty(s)) continue;
```
Comment (cornelissenja) [inline WebClient/src/components/moduleControl/moduleModel.js:21]: It took me a while to understand this function checks if s is a non-inherited property so only the real statuses are checked, maybe a little extra info?

File: WebClient/src/components/moduleControl/moduleControlWindow/moduleControlItemList/moduleControlItemListView.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        if (!opts.modules) throw new Error('opts.modules property is not provided');
+
+        this._moduleItems = {};
+        this._modules = opts.modules;
+    },
+
+    onRender: function () {
+        this._rootElement = L.DomUtil.create('div', 'module-control-item-list', this._parent);
+        
+        this._addModuleItems({ models: this._modules.models });
```
Comment (cornelissenja) [inline WebClient/src/components/moduleControl/moduleControlWindow/moduleControlItemList/moduleControlItemListView.js:24]: modules.models looks a bit inconsistent.
Please disregard , i get it now..

File: WebClient/src/core/modelCollection.js [hunk 8]
```diff
@@ -39,18 +39,21 @@ var ModelCollection = L.Evented.extend({
...
             var removedModels = difference(models, newModels);
+            var addedModels = difference(newModels, models);
             models = toArray(newModels);
             
             this.fire('remove', { models: removedModels });
+            this.fire('add', { models: addedModels });            
             this.fire('change', { models: models });
         };
 
         this.add = function (newModels) {
```
Comment (cornelissenja) [inline WebClient/src/core/modelCollection.js:50]: Propose rename: newModels -> addModels, it is not the complete list?

File: WebClient/src/components/moduleControl/moduleModel.js (hunk 6 above)
Comment (korenkovv) [reply WebClient/src/components/moduleControl/moduleModel.js:21]: This is a generally used check when using "for in" loop to iterate through an object.
The problem is that "for in" will go through every property on the prototype chain, including ones that we don't know about. 
So it is recommended to always use "hasOwnProperty" check with "for in" loop unless you need to iterate through all of the properties.

File: WebClient/src/core/modelCollection.js (hunk 8 above)
Comment (korenkovv) [reply WebClient/src/core/modelCollection.js:50]: in this case: 
- newModels is a new list of models that should override models field.
- addedModels is a list of models that were introduced by the newModels list and were not present in the models list.

**Review by cornelissenja — APPROVED** [review]
Go!?

**Pull request merged**